### ISSUE TITLE
Feature request: mapMaybe

### ISSUE THREAD

**Issue by GregorySchwartz** [Gabriel439/Haskell-Turtle-Library#366]
I find myself doing `fromJust . filter isJust . fmap f`, where `f :: a -> Maybe b`. Is it possible to have a `mapMaybe`-like function for this to avoid importing and using partial functions?

**Comment by Gabriel439** [comment]
@GregorySchwartz: What is the type signature that you expect for this utility?

**Comment by GregorySchwartz** [comment]
`(a -> Maybe b) -> Shell a -> Shell b`. Or, preferably, have an instance of Shell that would accomplish the same task (and more through generalization).

**Comment by int-index** [comment]
Does this do what you want?

```
filterShell :: (a -> Maybe b) -> Shell a -> Shell b
filterShell p s = fmap p s >>= select
```

**Comment by Gabriel439** [comment]
It seems like this doesn't have to be `turtle`-specific.  This utility could be generalized to any `MonadPlus`:

```haskell
example :: MonadPlus m => (a -> Maybe a) -> m a -> m a
example f m = do
    a <- m

    case f a of
        Nothing -> mzero
        Just b  -> return b
```

I'm actually surprised that there's not already something like this in `Control.Monad`.

**Comment by miguel-negrao** [comment]
Wouldn't the more general function be the following ?

```Haskell
example2 :: MonadPlus m => m (Maybe a) -> m a
example2 m = do
    a <- m
    case a of
        Nothing -> mzero
        Just b  -> return b
```

Then `example f m = example2 $ fmap f m`

**Comment by GregorySchwartz** [comment]
Yes, that would be perfect (although simplier than I thought it would be, which is always a good thing!).

**Comment by Gabriel439** [comment]
@GregorySchwartz: Could you try to see if this could be added to `base`?  If they turn it down, though, then I can add it to `turtle`

**Comment by hseg** [comment]
This exists as the `mapMaybe` method of `Filterable` in `witherable`.
It seems the `WrappedFoldable` instance is the one wanted here (perhaps via `DerivingVia`?):
```haskell
mapMaybe :: (Foldable f, Alterative f) => (a -> Maybe b) -> f a -> f b
mapMaybe f = foldr (\x xs -> case p x of {Just a -> pure a <|> xs; _ -> xs}) empty
```
